sky/provision/oci/query_utils.py

```diff
@@ -468,5 +472,194 @@ def delete_vcn(cls, net_client, skypilot_vcn, skypilot_subnet,
+                compartment_id=compartment,
+                vcn_id=vcn.id,
+                display_name=f'nsg_{cluster_name}',
+            ))
+        get_nsg_resp = net_client.get_network_security_group(
+            network_security_group_id=create_nsg_resp.data.id)
+        oci_adaptor.oci.wait_until(
+            net_client,
+            get_nsg_resp,
+            'lifecycle_state',
+            'AVAILABLE',
+        )
+
+        return get_nsg_resp.data.id
+
+    @classmethod
+    @debug_enabled(logger)
+    def add_ingress_rules(cls, region: str, cluster_name: str,
+                          ports: Set[int]) -> None:
+        if not ports:
+            return
+
+        net_client = oci_adaptor.get_net_client(
+            region, oci_utils.oci_config.get_profile())
+
+        nsg_id = cls.find_or_create_nsg(region, cluster_name)
+
+        filters = {constants.TAG_RAY_CLUSTER_NAME: cluster_name}
+        insts = query_helper.query_instances_by_tags(filters, region)
+        for inst in insts:
+            vnic = cls.get_instance_primary_vnic(
+                region=region,
+                inst_info={
+                    'inst_id': inst.identifier,
+                    'ad': inst.availability_domain,
+                    'compartment': inst.compartment_id,
+                })
+            nsg_ids = vnic.nsg_ids
+            if not nsg_ids:
+                net_client.update_vnic(
+                    vnic_id=vnic.id,
+                    update_vnic_details=oci_adaptor.oci.core.models.
+                    UpdateVnicDetails(nsg_ids=[nsg_id],
+                                      skip_source_dest_check=False),
+                )
+
+        # pylint: disable=line-too-long
+        list_nsg_rules_resp = net_client.list_network_security_group_security_rules(
+            network_security_group_id=nsg_id,
+            direction='INGRESS',
+            sort_by='TIMECREATED',
+            sort_order='DESC',
+        )
+
+        ingress_rules: List = [
+            r for r in list_nsg_rules_resp.data if r.direction == 'INGRESS'
+        ]
+
+        existing_ports: Set[int] = set()
+
+        for r in ingress_rules:
+            rule_existing_ports = [
+                p for p in ports if r.tcp_options and
+                (r.tcp_options.destination_port_range.max >= p and
+                 r.tcp_options.destination_port_range.min <= p)
+            ]
+            existing_ports.update(rule_existing_ports)
+
+        new_ports = ports - existing_ports
+        new_rules = [
+            oci_adaptor.oci.core.models.AddSecurityRuleDetails(
+                direction='INGRESS',
+                protocol='6',
+                is_stateless=False,
+                source=oci_utils.oci_config.VCN_CIDR_INTERNET,
+                source_type='CIDR_BLOCK',
+                tcp_options=oci_adaptor.oci.core.models.TcpOptions(
+                    destination_port_range=oci_adaptor.oci.core.models.
+                    PortRange(max=p, min=p),),
```

Comment:
IIUC this will create 100 rules for ranges like 30000-30100. Should we create rules based on range? You could check this function:

https://github.com/skypilot-org/skypilot/blob/b8416b8434dd1f2e70e2514d3cac8e7698b32088/sky/utils/resources_utils.py#L106-L129

Reply:
After this modification, the delete security rules function would become a little bit tricky, but fortunately we dont need such fine-grained API for now - we can directly cleanup the whole nsg at cluster termination.

Reply:
My implementation considerred the case where incrementally add/delete ports or user manually editted the nsg to add their ports. It is simpler if we can remove it as a whole. I'll change it to deal the nsg as a whole.

Reply:
Modified